packages/node-runtime-worker-thread/src/serializer.ts

```diff
@@ -0,0 +1,145 @@
+  serializedValue: Document;
+  nonEnumPropertyDescriptors: Record<string, PropertyDescriptor>;
+};
+
+export function serializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  // Type `null` indicates anything that is not returned by shell-api, any
+  // usual javascript value returned from evaluation
+  if (result.type === null) {
+    // Errors get special treatment to achieve better serialization
+    if (isError(result.rawValue)) {
+      result.type = SerializedResultTypes.SerializedErrorResult;
+      result.printable = serializeError(result.rawValue);
+    } else if (!isPrimitive(result.rawValue)) {
+      // For everything else that is not a primitive value there is just too
+      // many possible combinations of data types to distinguish between them,
+      // if we don't know the type and it's not an error, let's inspect and
+      // return an inspection result
+      result.type = SerializedResultTypes.InspectResult;
+      result.printable = inspect(result.rawValue);
+    }
+  } else if (!isPrimitive(result.printable)) {
+    // If type is present we are dealing with shell-api return value. In most
+    // cases printable values of shell-api are primitive or serializable as
+    // (E)JSON, but they also might have some non-enumerable stuff (looking at
+    // you, Cursor!) that we need to serialize additionally, otherwiser those
+    // params will be lost in the ether causing issues down the way
+    const origType = result.type;
+    result.type = SerializedResultTypes.SerializedShellApiResult;
+    result.printable = {
+      origType,
+      serializedValue: EJSON.serialize(result.printable),
+      nonEnumPropertyDescriptors: Object.fromEntries(
+        Object.entries(
+          Object.getOwnPropertyDescriptors(result.printable)
+        ).filter(([, descriptor]) => !descriptor.enumerable)
+      )
+    };
+  }
+
+  // `rawValue` can't be serialized "by design", we don't really care about it
+  // for the worker thread runtime so as an easy workaround we will just remove
+  // it completely from the result
+  delete result.rawValue;
+
+  return result;
+}
+
+export function deserializeEvaluationResult(result: ShellResult): ShellResult {
+  result = { ...result };
+
+  if (result.type === SerializedResultTypes.SerializedErrorResult) {
+    result.type = null;
+    result.printable = deserializeError(result.printable);
+  }
+
+  if (result.type === SerializedResultTypes.SerializedShellApiResult) {
+    const value: SerializedShellApiResult = result.printable;
+    const printable = EJSON.deserialize(value.serializedValue);
+
+    // Primitives should not end up here ever, but we need to convince TS that
+    // its true
+    if (!isPrimitive(printable)) {
+      Object.defineProperties(printable, value.nonEnumPropertyDescriptors);
+    }
+
+    result.type = value.origType;
+    result.printable = printable;
+  }
+
+  Object.defineProperty(result, 'rawValue', {
+    get() {
```

Comment:
Why do we have this here? Why we don't just remove `rawValue` from any typescript interface used by consumers and remove this code

Reply:
Consumers mostly rely on interfaces exposed by shell-api, which are pretty generic to match all possible runtimes, because of that not having `rawValue` breaks type assertions. Changing it for `serializer` will cause a chain effect where `WorkerRuntime` can't be an implemenation of `Runtime` anymore and all the clients using it can't use Runtimes if they are using WorkerRuntime, and so on. In my opinion having this runtime guard against using `rawValue` is a good compromise to this. What do you think?

Reply:
Decided f2f that runtime evaluation results should not include `rawValue`. Done in 9acceee